tests/e2e/hashkv_test.go

```diff
@@ -0,0 +1,359 @@
+			hashKVOnTombstoneRev:    false,
+		},
+		{
+			compactedOnTombstoneRev: false,
+			hashKVOnTombstoneRev:    false,
+		},
+	}
+
+	scenarios := []struct {
+		ClusterVersion fe2e.ClusterVersion
+		OnlyOneKey     bool
+	}{
+		{
+			ClusterVersion: fe2e.CurrentVersion,
+			OnlyOneKey:     true,
+		},
+		{
+			ClusterVersion: fe2e.CurrentVersion,
+			OnlyOneKey:     false,
+		},
+		{
+			ClusterVersion: fe2e.QuorumLastVersion,
+			OnlyOneKey:     true,
+		},
+		{
+			ClusterVersion: fe2e.QuorumLastVersion,
+			OnlyOneKey:     false,
+		},
+	}
+
+	for idx, tt := range tests {
+		for _, scenario := range scenarios {
+			t.Run(fmt.Sprintf("#%d - %s - OnlyOneKey=%v", idx, scenario.ClusterVersion, scenario.OnlyOneKey), func(t *testing.T) {
+				ctx := context.Background()
+
+				clus, cliCfg := newClusterForHashKV(t, scenario.ClusterVersion)
+
+				dataset := newTestDatasetInCluster(t, clus, cliCfg, scenario.OnlyOneKey)
+
+				// ensure we can pick up the right revision
+				require.True(t, len(dataset.tombstones) > 2)
+				require.True(t, dataset.tombstones[0]+3 < dataset.tombstones[1])
+				require.True(t, dataset.tombstones[1]+3 < dataset.tombstones[2])
+
+				compactedOnRev := dataset.tombstones[0]
+				if !tt.compactedOnTombstoneRev {
+					compactedOnRev += 3
+				}
+
+				hashKVOnRev := dataset.tombstones[1]
+				if !tt.hashKVOnTombstoneRev {
+					hashKVOnRev += 3
+				}
+
+				cli, err := fe2e.NewEtcdctl(cliCfg, clus.EndpointsGRPC())
+				require.NoError(t, err)
+
+				t.Logf("COMPACT on rev=%d", compactedOnRev)
+				_, err = cli.Compact(ctx, compactedOnRev, config.CompactOption{Physical: true})
+				require.NoError(t, err)
+
+				t.Logf("HashKV on rev=%d", hashKVOnRev)
+				resp, err := cli.HashKV(ctx, hashKVOnRev)
+				require.NoError(t, err)
+
+				require.Len(t, resp, 3)
+				require.True(t, resp[0].Hash != 0)
+				t.Logf("One Hash value is %d", resp[0].Hash)
+
+				require.Equal(t, resp[0].Hash, resp[1].Hash)
+				require.Equal(t, resp[1].Hash, resp[2].Hash)
+			})
+		}
+	}
+}
+
+type datasetInfo struct {
+	keys map[string]struct{}
+
+	tombstones      []int64
+	lastestRevision int64
+}
+
+// key: "foo"
+// modified: 41
+// generations:
+//
+//	{34, 0}, {35, 0}, {36, 0}, {37, 0}, {38, 0}, {39, 0}, {40, 0}, {41, 0}
+//	{23, 0}, {24, 0}, {25, 0}, {26, 0}, {27, 0}, {28, 0}, {29, 0}, {30, 0}, {31, 0}, {32, 0}, {33, 0}[tombstone]
+//	{12, 0}, {13, 0}, {14, 0}, {15, 0}, {16, 0}, {17, 0}, {18, 0}, {19, 0}, {20, 0}, {21, 0}, {22, 0}[tombstone]
+//	{2, 0},  {3, 0},  {4, 0},  {5, 0},  {6, 0},  {7, 0},  {8, 0},  {9, 0},  {10, 0}, {11, 0}[tombstone]
+//
+// # If onlyOneKey is false
+//
+// key: "foo-1"
+// modified: 12
+// generations:
+//
+//	{12, 1}
+//
+// key: "foo-2"
+// modified: 23
+// generations:
+//
+//	{23, 1}
+//
+// key: "foo-3"
+// modified: 34
+// generations:
+//
+//	{34, 1}
+func newTestDatasetInCluster(t *testing.T, clus *fe2e.EtcdProcessCluster, cliCfg fe2e.ClientConfig, onlyOneKey bool) *datasetInfo {
```

Comment:
It's unnecessary complicated.

Please consider a function something like below,
```
// populateData populates some sample data, and return a slice of tombstone revisions and the latest revision
func populateData(t *testing.T, clus *fe2e.EtcdProcessCluster, clientCfg fe2e.ClientConfig, keys []string) ([]int64, int64) {
	c := newClient(t, clus.EndpointsGRPC(), clientCfg)

	ctx := context.Background()
	totalOperations := 40

	var (
		tombStoneRevs []int64
		latestRev     int64
	)

	deleteStep := 10 // submit a delete operation on every 10 operations
	for i := 1; i <= totalOperations; i++ {
		if i%deleteStep == 0 {
			t.Logf("Deleting key=%s", keys[0]) // Only delete the first key for simplicity
			resp, derr := c.Delete(ctx, keys[0])
			require.NoError(t, derr)
			latestRev = resp.Header.Revision
			tombStoneRevs = append(tombStoneRevs, resp.Header.Revision)
			continue
		}

		value := fmt.Sprintf("%d", i)
		var ops []clientv3.Op
		for _, key := range keys {
			ops = append(ops, clientv3.OpPut(key, value))
		}

		t.Logf("Writing keys: %v, value: %s", keys, value)
		resp, terr := c.Txn(ctx).Then(ops...).Commit()
		require.NoError(t, terr)
		require.True(t, resp.Succeeded)
		require.Len(t, resp.Responses, len(ops))
		latestRev = resp.Header.Revision
	}

	return tombStoneRevs, latestRev
}
```

Reply:
Thanks. I used your code in latest commit. Please take a look.